src/stream/src/executor/gap_fill.rs

```diff
@@ -0,0 +1,1856 @@
+        self.state_table.init_epoch(epoch).await
+    }
+
+    pub fn insert(&mut self, value: impl Row) {
+        self.state_table.insert(value);
+    }
+
+    pub fn delete(&mut self, value: impl Row) {
+        self.state_table.delete(value);
+    }
+
+    /// Batch delete multiple rows from state table.
+    pub fn batch_delete(&mut self, rows: Vec<impl Row>) {
+        for row in rows {
+            self.state_table.delete(row);
+        }
+    }
+
+    /// Scans `StateStore` for filled rows between two time points (exclusive).
+    pub async fn scan_filled_rows_between(
+        &self,
+        start_time: &GapFillCacheKey,
+        end_time: &GapFillCacheKey,
+    ) -> StreamExecutorResult<Vec<(OwnedRow, OwnedRow)>> {
+        let mut filled_rows_to_delete = Vec::new();
+
+        let start_time_row = self.time_key_serde.deserialize(start_time)?;
+        let end_time_row = self.time_key_serde.deserialize(end_time)?;
+
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) = &(
+            Bound::Excluded(start_time_row),
+            Bound::Excluded(end_time_row),
+        );
+
+        let state_table_iter = self
+            .state_table
+            .iter_with_prefix(
+                &None::<row::Empty>,
+                sub_range,
+                PrefetchOptions::prefetch_for_large_range_scan(),
+            )
+            .await?;
+
+        pin_mut!(state_table_iter);
+        while let Some(row_result) = state_table_iter.next().await {
+            let row = row_result?.into_owned_row();
+
+            if let Some(is_filled_datum) = row.datum_at(row.len() - 1)
+                && let Some(ScalarImpl::Bool(is_filled)) = is_filled_datum.to_owned_datum()
+                && is_filled
+            {
+                // This is a filled row, add to deletion list.
+                // For state table, we use the time column as the key.
+                let time_datum = row.datum_at(self.time_column_index);
+                let state_key = OwnedRow::new(vec![time_datum.to_owned_datum()]);
+                filled_rows_to_delete.push((state_key, row));
+            }
+        }
+
+        Ok(filled_rows_to_delete)
+    }
+
+    /// Serialize time value to cache key.
+    pub fn serialize_time_to_cache_key(&self, row: impl Row) -> GapFillCacheKey {
+        let time_datum = row.datum_at(self.time_column_index);
+        let time_row = OwnedRow::new(vec![time_datum.to_owned_datum()]);
+        time_row.memcmp_serialize(&self.time_key_serde)
+    }
+
+    /// Scans a range of data from `StateStore` to build a window of rows before `end_time`.
+    pub async fn scan_range_before(
+        &self,
+        end_time: &GapFillCacheKey,
+        limit: usize,
+    ) -> StreamExecutorResult<Vec<(GapFillCacheKey, CompactedRow, RowType)>> {
+        let end_time_row = self.time_key_serde.deserialize(end_time)?;
+
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) =
+            &(Bound::Unbounded, Bound::Excluded(end_time_row));
```

Comment:
This means we scan all the way from the very start. I don't think this is ideal. Can you use the state table's reverse iterator instead?

```
    pub async fn rev_iter_with_prefix(
        &self,
        pk_prefix: impl Row,
        sub_range: &(Bound<impl Row>, Bound<impl Row>),
        prefetch_options: PrefetchOptions,
    ) -> StreamExecutorResult<impl RowStream<'_>> {
        Ok(
            self.iter_with_prefix_inner::</* REVERSE */ true, ()>(pk_prefix, sub_range, prefetch_options)
                .await?.map_ok(|(_, row)| row),
        )
    }
```